tests/test_runner_apps/circular_ref/models.py

```diff
@@ -0,0 +1,39 @@
+from django.db import models
+
+
+class ByNameManager(models.Manager):
+    def get_by_natural_key(self, key):
+            return self.get(key=key)
+
+
+# Module with a circular reference through CircularRefB, and with natural keys
+class CircularRefThingA(models.Model):
+    key = models.CharField(max_length=100)
+    other_thing = models.ForeignKey('CircularRefThingB', on_delete=models.CASCADE, null=True)
+
+    objects = ByNameManager()
+
+    def natural_key(self):
+        return (self.key,)
+
+
+# Module with a circular reference through CircularRefA, and with natural keys
+class CircularRefThingB(models.Model):
+    key = models.CharField(max_length=100)
+    other_thing = models.ForeignKey('CircularRefThingA', on_delete=models.CASCADE, null=True)
+
+    objects = ByNameManager()
+
+    def natural_key(self):
+        return (self.key,)
+
+
+# Module containing a self-reference, and with natural keys
+class SelfRefThing(models.Model):
+    key = models.CharField(max_length=100)
```

Comment:
`key` should likely be unique if it's used as a natural key?

Reply:
Good point. I based this off an existing model that also does not specify uniqueness: https://github.com/django/django/blob/c33eb6dcd0c211f8f02b2976fe3b3463f0a54498/tests/fixtures/models.py#L121-L124

Enforcing uniqueness is not relevant to the testcase (that is, things are unique in the testdata, which is sufficient), so keeping it like this might be more consistent. Alternatively, I can make the existing model also unique, that might actually be a good idea.

Reply:
Done, I made changed the existing model touched by the commit as well. There's still a lot of other natural key models without uniqueness in other tests, but I think that is out of scope, so I'm resolving this. I will push the changes soon.

Reply:
I cleaned it up in #12653.